File: src/c++/0-99/35_search_insert_position.cpp
/**
 * Solution to 35. Search Insert Position (https://leetcode.com/problems/search-insert-position/)
 * Difficulty: EASY
 * Runtime: O(log n)
 * Submission: https://leetcode.com/submissions/detail/597709327/
 *
 * @author jagwirecode
*/

class Solution
{
public:
    int searchInsert(vector<int> &nums, int target)
    {
        int size = nums.size();

        // Bounds check
        if (size < 1)
        {
            return 0;
        }

        int low = 0;
        int high = size - 1;
        int mid;
        int index = 0;

        // Begin binary search
        while (low <= high)
        {
            // Safely calculate midpoint
            mid = low + ((high - low) / 2);

            // Lower half check - update where index should be if not present (always 1 more than final midpoint)
            if (nums.at(mid) < target)
            {
                low = mid + 1;
                index = mid + 1;
            }
            // Upper half check
            else if (nums.at(mid) > target)
            {
                high = mid - 1;
            }

            // Element found
            else
            {
                return mid;
            }
        }

        // Return where it should have been if it were present
        return index;
    }
};

File: src/python/0-99/35_search_insert_position.py
"""
Solution to 35. Search Insert Position (https://leetcode.com/problems/search-insert-position/)
Difficulty:     EASY
Runtime:        O(log n) - Logarithmic time
Space:          O(1) - Constant space
Submission:     https://leetcode.com/submissions/detail/605862436/
"""
__author__      = "jagwirecode"

from typing import List

class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        """
        Searches for the target, returns the index if found, otherwise returns the index of where
        the target should be if it were inserted.

        Args:
            nums (List[int]): List of distinct, sorted ascendingly, integers
            target [int]: Number who's index is to be identified
        
        Returns:
            [int]: Index of the number in the list, or where it would be if not currently present
        """    
        low = 0
        high = len(nums) - 1
        mid = 0
        index = 0
        
        # Binary search method
        while(low <= high):
            mid = low + ((high - low) // 2)
            
            if(nums[mid] < target):
                low = index = mid + 1
            elif(nums[mid] > target):
                high = mid - 1
            else:
                return mid
            
        return index
     